README.md

```diff
@@ -56,7 +56,28 @@ It has specific TDR commands around realm import, and is copied from the default
 
 The standalone-ha.xml is mostly the standard configuration for keycloak with a few changes to get it to work with the load balancer. Some of these are discussed in the keycloak [documentation](https://www.keycloak.org/docs/latest/server_installation/#_setting-up-a-load-balancer-or-proxy)
 
-## Updating Keycloak Configuration json
+## Updating TDR Realm Configuration json
+
+A separate Jenkins job is used to update the TDR realm configuration: TDR Auth Server Update
+
+This is because on container start up, if a realm already exists, its configuration json is ignored. So restarting the container will not update the configuration.
+
+The TDR Auth Server Update Jenkins job updates existing realm configuration using the Keycloak REST APIs: https://www.keycloak.org/docs-api/11.0/rest-api/index.html
+
+Two REST endpoints are used to update a Keycloak realm:
+* `PUT /{realm}`: updates the top-level information of the realm.    
+  * Takes the full realm representation: https://www.keycloak.org/docs-api/11.0/rest-api/index.html#_realmrepresentation
+  * Any user, roles or client information in the representation are ignored.
+* `POST /{realm}/partialImport`: Partial import from a JSON file to an existing realm.
+  * Takes a partial realm representation: https://www.keycloak.org/docs-api/11.0/rest-api/index.html#_partialimportrepresentation
+  * Allows importation of configuration for: `clients`, `roles`, `groups` and `identity providers` only
+  * Uses a policy option that controls what behaviour occurs for existing realm resources:
+    * `SKIP`: Skips any existing resources and does not update them
```

Comment:
Does this mean we have no way of updating properties of existing resources, like group permissions?

Reply:
I think yes, you wouldn't be able to update existing resources. However, I can't be 100% sure of the exact behaviour of Keycloak, without explicitly trying it out.

Reply:
So what kind of operation do you think we'll have to do manually?

Reply:
I would assume any operation that would change the JSON configuration within the structure of an existing resource.

Reply:
What sort of thing? It sounds like associating a role with a group will need to be done manually - is that right? What else?

I'm just trying to work out how hard the maintenance will be in practice.

Reply:
Based on the JSON structure the following elements would not be updated in an existing resource:
**Group**
- attributes
- realmRoles
- clientRoles
- subGroups

**Client**
- clientId
- surrogateAuthRequired
- enabled
- alwaysDisplayInConsole
- clientAuthenticatorType
- secret
- redirectUris
- webOrigins
- notBefore
- bearerOnly
- consentRequired
- standardFlowEnabled
- implicitFlowEnabled
- directAccessGrantsEnabled
- serviceAccountsEnabled
- publicClient
- frontchannelLogout
- protocol
- attributes
- authenticationFlowBindingOverrides
- fullScopeAllowed
- nodeReRegistrationTimeout
- protocolMappers
- defaultClientScopes
- optionalClientScopes

**Roles**
- id
- name
- description
- composite
- clientRole
- containerId
- attributes

**Identity Providers**
- addReadTokenRoleOnCreate
- alias
- config
- displayName
- enabled
- firstBrokerLoginFlowAlias
- internalId
- linkOnly
- postBrokerLoginFlowAlias
- providerId
- storeToken
- trustEmail

Reply:
That's quite a lot! Lots of the client properties are things that would be safer in configuration-as-code. 😢 

Can you _get_ all of that information from the API? What do you think about adding a card about running a regular scan to compare the configuration with the expected configuration so that at least we can detect misconfiguration even if we can't set it automatically?

Reply:
Yes, you should be able to export the full config JSON via the API. The trick will be around the IDs in the config, as they'll be different for the different environments, so the comparison might be a little tricky.

Reply:
I've created https://national-archives.atlassian.net/browse/TDR-456 with a link to this discussion. 👍